### DISCUSSION TITLE
Gateway timeout uploading XML on sample proxy configuration

### DISCUSSION THREAD

**Discussion by melissaboiko** [dani-garcia/vaultwarden#1895] [Q&A]
v1.22.2, fresh install, exposed via nginx reverse proxy via configuration from the wiki https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples . The host nginx is 1.14.2, on Debian buster.

When I try to import my Keepass data (at ~2MiB XML), it hangs for a while, then raises an error with nginx giving a 504 proxy timeout.  It appears to have nonetheless imported all the data.

I tried increasing the proxy wait time as follows:

```nginx
server {
  # ...
  proxy_connect_timeout       777;
  proxy_send_timeout          777;
  proxy_read_timeout          777;
  send_timeout                777;
```

and this appears to have been enough, when I looked at the tab again it was back to the main vault page.  However, because I didn't realise it was importing the data despite the 504, I ended up with several duplicates of each password.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.22.2
* Web-vault version: v2.21.1
* Running within Docker: true
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: false
* HTTPS Check: false
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "****://*********",
  "domain_origin": "****://*********",
  "domain_path": "",
  "domain_set": false,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": null,
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Deployment environment

* vaultwarden version: v1.22.2,

* Install method: docker run -d --name bitwarden \
              -e ADMIN_TOKEN=[token] \
              -v /vw-data/:/data/ \
              -p 40080:80 \
              vaultwarden/server:latest

* Clients used: firefox

* Reverse proxy and version: nginx 1.14.2

* MySQL/MariaDB or PostgreSQL version: 

* Other relevant details: Host has 8GiB RAM, amd64 4 cores 3600Mhz and is not under load.

**Comment by BlackDex** [comment]
A 504 is a proxy/gateway timeout, which indicates that its an (configuration) issue of nginx.

Nginx waited for vaultwarden to reply back after the upload, which took a long time, and in the end nginx gave up, but Vaultwarden continued and processed the incoming data.

**Reply by melissaboiko** [reply]
The issue is that nginx will do that with the configuration provided in the wiki.  So if it's normal for vaultwarden to take longer than the default proxy timeout, it seems like it would make sense for the provided nginx proxy configuration to include longer proxy wait parameters.

**Reply by BlackDex** [reply]
Well, that is just default nginx config. Every reverse proxy has different settings, so going to process them all is probably not very useful.

Though, a mention of this somewhere probably doesn't harm. Since it's a wiki, feel free to help and optimize.

But, i also imported a large keepass, and i had no issues. But i have it running on a powerful server. So it will be different for everybody, and it also depends on the database backend, storage etc...
